lib/pure/collections/hashcommon.nim

```diff
@@ -20,32 +20,66 @@ when not defined(nimHasDefault):
 
 # hcode for real keys cannot be zero.  hcode==0 signifies an empty slot.  These
 # two procs retain clarity of that encoding without the space cost of an enum.
-proc isEmpty(hcode: Hash): bool {.inline.} =
-  result = hcode == 0
+when false:
+  proc isEmpty(hcode: Hash): bool {.inline.} =
+    result = hcode == 0
+
+const freeMarker = 0
+const deletedMarker = -1
+
+proc isFilledValid(hcode: Hash): bool {.inline.} =
+  result = hcode != 0 and hcode != deletedMarker # SPEED: could improve w bit magic
 
 proc isFilled(hcode: Hash): bool {.inline.} =
   result = hcode != 0
 
-proc nextTry(h, maxHash: Hash): Hash {.inline.} =
-  result = (h + 1) and maxHash
+
+proc nextTry(h, maxHash: Hash, perturb: var Hash): Hash {.inline.} =
+  # TODO: shouldn't perturb be unsigned?
+  const PERTURB_SHIFT = 5
+  # SPEED IMPROVE
```

Comment:
The technical term is `performance`. I know you make this mistake regularly, so please write it down.